Slugify Collectives and Pages

Koc opened the pull request:
### 📝 Summary

This is a very beginning of my PR to slugify urls for Collectives and Pages. For now urls are too long and hard to understand if using cyrylic in this entities.

Here an example of some of them: `http://localhost/index.php/apps/collectives/%D0%A3%D0%BA%D1%80%D0%B0%D1%97%D0%BD%D1%81%D1%8C%D0%BA%D1%96%20%D1%84%D1%83%D1%82%D0%B1%D0%BE%D0%BB%D1%96%D1%81%D1%82%D0%B8/%D0%AF%D1%80%D0%BC%D0%BE%D0%BB%D0%B5%D0%BD%D0%BA%D0%BE%20%D0%90%D0%BD%D0%B4%D1%80%D1%96%D0%B9%20%D0%9C%D0%B8%D0%BA%D0%BE%D0%BB%D0%B0%D0%B9%D0%BE%D0%B2%D0%B8%D1%87?fileId=187`. Just compare it with slugified version: `http://localhost/index-php/apps/collectives/ukrayinski-futbolisti-1/page-1-yurchenko-andriy-mykolayovich`. Easy to read, 3 times shorter.


#### 🖼️ Video
[Screencast from 2024-09-01 17-18-46.webm](https://github.com/user-attachments/assets/47e96118-8425-4926-bcbd-605f7042740d)


### 🚧 TODO
- [x] resolve updater [issue](https://github.com/nextcloud/collectives/pull/1424#discussion_r1819174500)
- [x] update previewer
- [x] add/fix tests

### 🏁 Checklist

- [x] Code is properly formatted (`npm run lint` / `npm run stylelint` / `composer run cs:check`)
- [x] [Sign-off message](https://probot.github.io/apps/dco/) is added to all commits
- [ ] Tests (unit, integration and/or end-to-end) passing and the changes are covered with tests
- [ ] Documentation ([README](https://github.com/nextcloud/collectives/blob/main/README.md) or [documentation](https://github.com/nextcloud/collectives/blob/main/docs/)) has been updated or is not required

mejo- commented:
Hey @Koc, thanks for looking into this. Definitely a good idea to slugify collectives and page titles for cleaner URLs.

Without looking into all the details, I wonder if we could implement this in a way that doesn't need the sluggified string to be persisted in the database.

Why not make calculate the slug of titles/names on demand in the backend and add them as attribute to the collectives and pages in the API responses? Then the frontend could always check for name/title match first and for slug match second. What do you think about this approach?

Koc commented:
@mejo- we have a lot of pages in collectives, so I would like to reduce amount of runtime operations to speedup backend endpoints.

Koc commented:
I've achieved some progress. Now it more or less works, we have even redirects if slugs have been renamed

Koc commented:
@mejo- there is one more reason to persist `slug` in the database: we can make it editable in upcoming PRs. Sometimes page titles can be long, so we can give possibility to our users shorten them

Koc commented:
For some reason migrations stop work for me (even previous). Both autoloading and dependency injection resolving not works. Any idea how to fix that?

```
devcontainer@5e275b290d01:/var/www/html$ ./occ migrations:migrate collectives -v

In MigrationService.php line 399:
                                                                                         
  [Exception]                                                                            
  Database error when running migration 021200Date20240725154232 for app collectives     
  Migration step 'OCA\Collectives\Migration\Version021200Date20240725154232' is unknown  
                                                                                         

Exception trace:
  at /var/www/html/lib/private/DB/MigrationService.php:399
 OC\DB\MigrationService->migrate() at /var/www/html/core/Command/Db/Migrations/MigrateCommand.php:42
 OC\Core\Command\Db\Migrations\MigrateCommand->execute() at /var/www/html/3rdparty/symfony/console/Command/Command.php:298
 Symfony\Component\Console\Command\Command->run() at /var/www/html/3rdparty/symfony/console/Application.php:1040
 Symfony\Component\Console\Application->doRunCommand() at /var/www/html/3rdparty/symfony/console/Application.php:301
 Symfony\Component\Console\Application->doRun() at /var/www/html/3rdparty/symfony/console/Application.php:171
 Symfony\Component\Console\Application->run() at /var/www/html/lib/private/Console/Application.php:187
 OC\Console\Application->run() at /var/www/html/console.php:87
 require_once() at /var/www/html/occ:11

In MigrationService.php line 482:
                                                                                         
  [InvalidArgumentException]                                                             
  Migration step 'OCA\Collectives\Migration\Version021200Date20240725154232' is unknown  
                                                                                         

Exception trace:
  at /var/www/html/lib/private/DB/MigrationService.php:482
 OC\DB\MigrationService->createInstance() at /var/www/html/lib/private/DB/MigrationService.php:497
 OC\DB\MigrationService->executeStep() at /var/www/html/lib/private/DB/MigrationService.php:395
 OC\DB\MigrationService->migrate() at /var/www/html/core/Command/Db/Migrations/MigrateCommand.php:42
 OC\Core\Command\Db\Migrations\MigrateCommand->execute() at /var/www/html/3rdparty/symfony/console/Command/Command.php:298
 Symfony\Component\Console\Command\Command->run() at /var/www/html/3rdparty/symfony/console/Application.php:1040
 Symfony\Component\Console\Application->doRunCommand() at /var/www/html/3rdparty/symfony/console/Application.php:301
 Symfony\Component\Console\Application->doRun() at /var/www/html/3rdparty/symfony/console/Application.php:171
 Symfony\Component\Console\Application->run() at /var/www/html/lib/private/Console/Application.php:187
 OC\Console\Application->run() at /var/www/html/console.php:87
 require_once() at /var/www/html/occ:11

migrations:migrate <app> [<version>]
```

mejo- commented on lib/Migration/Version021200Date20240820000001.php:
```diff
@@ -0,0 +1,94 @@
+		while ($rowCollective = $resultCollectives->fetch()) {
+			$circle = $this->circleHelper->getCircle($rowCollective['circle_unique_id'], null, true);
+			/** @var PageInfo[] $pageInfos */
+			$pageInfos = $this->service->findAll($rowCollective['id'], $circle->getOwner()->getUserId());
```
Unfortunately this won't work for trashed pages and neither for trashed collectives. I guess it's cleaner to write a dedicated function that unconditionally gets the filenames for all pages from `oc_collectives_pages` for a specific collective and iterates over it.

Please make sure you have trashed collectives *and* active collectives with trashed pages when testing it :wink:

Koc replied:
I will double check with trashed collectives/pages but IMHO this is fine™. Code should work properly even if slugs are not generated.

mejo- replied:
Well, with a collective in the collectives trash, the commented line throws a `NotFoundException` for me. This can be mitigated by adjusting the query to fetch collectives to only take into account non-trashed collectives:

```diff
--- a/lib/Migration/Version021500Date20240820000001.php
+++ b/lib/Migration/Version021500Date20240820000001.php
@@ -57,7 +57,9 @@ class Version021500Date20240820000001 extends SimpleMigrationStep {
                }
 
                $queryCollectives = $this->connection->getQueryBuilder();
-               $queryCollectives->select(['id', 'circle_unique_id'])->from('collectives');
+               $queryCollectives->select(['id', 'circle_unique_id'])
+                       ->from('collectives')
+                       ->where('trash_timestamp IS NULL');
                $resultCollectives = $queryCollectives->executeQuery();
 
                $queryPages = $this->connection->getQueryBuilder();
```

But still the commented line breaks in my tests with `NotFoundException` as the collective folder cannot be accessed - at least when I run the migration by calling `occ upgrade` on the commandline. I would have to further look into it, but I guess it's because the occ command doesn't run as a specific user, so the collective mountpoints are not accessible.

Could you check if the migration works smoothly if you have different users with collectives and pages and if you trigger the migration via running `occ upgrade`?

You can easily test the migrations again after the following SQL statements:

```sql
ALTER TABLE oc_collectives DROP COLUMN `slug`;
ALTER TABLE oc_collectives_pages DROP COLUMN `slug`;
DELETE FROM oc_migrations WHERE app='collectives' AND version='021500Date20240820000000';
DELETE FROM oc_migrations WHERE app='collectives' AND version='021500Date20240820000001';
```

Since the migration versions changed now, you might have to adjust this.

Afterwards just change the app version number in `appinfo/info.xml` (doesn't matter if you bump or lower it). This will make Nextcloud run the upgrade steps which will run all migrations that are not registered as already executed.

If I do this with pages in the trash, the commented line will throw a `NotFoundException`. This can be mitigated by the following code:

```php
try {
	$pageInfos = $this->service->findAll($rowCollective['id'], $circle->getOwner()->getUserId());
	} catch (NotFoundException
```

Koc replied:
I was able to reproduce an issue. But it is very strange :confused: .

So, for some reason `PageService::findAll` throws an exception during upgrade if at least 1 page in collective was trashed. But same method works fine in API.


```
string(29) "/admin/files/Collectives/aaaa"  <---- exception message with path + collective name

#0 /var/www/html/lib/private/Files/Node/Folder.php(107): OC\Files\Node\Root->get()
#1 /var/www/html/apps/collectives/lib/Service/PageService.php(100): OC\Files\Node\Folder->get() <--- why collective folder not available? Only 1 page was trashed, not whole collective
#2 /var/www/html/apps/collectives/lib/Service/PageService.php(498): OCA\Collectives\Service\PageService->getCollectiveFolder()
#3 /var/www/html/apps/collectives/lib/Migration/Version021500Date20240820000001.php(78): OCA\Collectives\Service\PageService->findAll()
#4 /var/www/html/lib/private/DB/MigrationService.php(522): OCA\Collectives\Migration\Version021500Date20240820000001->postSchemaChange()
#5 /var/www/html/lib/private/DB/MigrationService.php(395): OC\DB\MigrationService->executeStep()
#6 /var/www/html/lib/private/legacy/OC_App.php(694): OC\DB\MigrationService->migrate()
#7 /var/www/html/lib/private/Updater.php(325): OC_App::updateApp()
#8 /var/www/html/lib/private/Updater.php(236): OC\Updater->doAppUpgrade()
#9 /var/www/html/lib/private/Updater.php(100): OC\Updater->doUpgrade()
#10 /var/www/html/core/Command/Upgrade.php(190): OC\Updater->upgrade()
#11 /var/www/html/3rdparty/symfony/console/Command/Command.php(326): OC\Core\Command\Upgrade->execute()
#12 /var/www/html/3rdparty/symfony/console/Application.php(1078): Symfony\Component\Console\Command\Command->run()
#13 /var/www/html/3rdparty/symfony/console/Application.php(324): Symfony\Component\Console\Application->doRunCommand()
#14 /var/www/html/3rdparty/symfony/console/Application.php(175): Symfony\Component\Console\Application->doRun()
#15 /var/www/html/lib/private/Console/Application.php(187): Symfony\Component\Console\Application->run()
#16 /var/www/html/console.php(87): OC\Console\Application->run()
#17 /var/www/html/occ(11): require_once('...')
#18 {main}
```

mejo- commented on lib/Service/PageService.php:
```diff
@@ -1084,4 +1096,12 @@ public function getBacklinks(int $collectiveId, int $id, string $userId): array
+			return null;
+		}
+
+		return $this->slugGeneratorService->generatePageSlug($title);
```
I guess that you wanted to pass the `$fileId` here and add it to the sluggified title in `generatePageSlug`?

mejo- replied:
This still seems broken to me. The backend code doesn't add the page fileId to the slug, but the frontend code expects it to be there.

Koc replied:
no, it's not needed. We're matching `pageId` and `pageSlug` separately on FE. This needed because slug can change and we should redirect user to a new slug based on `pageId`

mejo- reviewed (changes requested):
Thanks for the PR, @Koc, much appreciated! Sorry for taking so long with a proper review.

I have a few comments and ran into a few problems testing it, but I also have s general topic that I would like to discuss within our team:

With the PR, Collectives URLs will always include collectiveId and pageId, e.g. `/apps/collectives/MyCollective/MyPage?fileId=123` will become `/apps/collectives/MyCollective-1/MyPage-123` in the future. I see the advantages and I don't have strong objections, but it's a user experience change and makes the URLs less readable for collectives and pages that only had short ascii names anyway. I think this is out-weighted by the fact that URLs become way cleaner by longer and non-ascii names. The IDs are required to prevent duplicates - and they also allow cleaner URL parsing in the frontend.

@juliusknorr @max-nextcloud @jancborchardt @nimishavijay what do you think about this?

jancborchardt reviewed (commented):
@Koc nice change, good for internationalization. :)

@mejo- I'm all for it too, the drawback is very very small it seems.

Koc commented on lib/Db/Page.php:
```diff
@@ -31,6 +33,7 @@
  */
 class Page extends Entity implements JsonSerializable {
 	protected ?int $fileId = null;
+	protected ?string $slug = null;
```
added

Koc commented on lib/Migration/Version021200Date20240820000001.php:
```diff
@@ -0,0 +1,94 @@
+	public function __construct(
+		private IDBConnection $connection,
+		private CircleHelper $circleHelper,
+		private PageService $service,
```
renamed

Koc commented on lib/Service/PageService.php:
```diff
@@ -325,7 +333,9 @@ private function newPage(int $collectiveId, Folder $folder, string $filename, st
 				$this->userManager->getDisplayName($userId));
-			$this->updatePage($collectiveId, $newFile->getId(), $userId);
+			$slug = $title ? $this->generateSlugForPage($title, $newFile) : null;
+			$this->updatePage($collectiveId, $newFile->getId(), $userId, slug: $slug);
```
replaced all introduced named parameters with a positional everywhere

Koc commented on lib/Service/PageService.php:
```diff
@@ -745,8 +755,9 @@ public function copy(int $collectiveId, int $id, ?int $parentId, ?string $title,
 		if (null !== $newFile = $this->moveOrCopyPage($collectiveFolder, $file, $parentId, $title, true)) {
 			$file = $newFile;
 		}
+		$slug = $title ? $this->generateSlugForPage($title, $file) : $this->generateSlugForPage($page->getTitle(), $file);
```
changed

Koc commented on lib/Service/SlugGeneratorService.php:
```diff
@@ -0,0 +1,27 @@
+
+use Symfony\Component\String\Slugger\SluggerInterface;
+
+class SlugGeneratorService {
```
renamed + all variables everywhere

Koc commented on lib/Service/PageService.php:
```diff
@@ -772,9 +783,10 @@ public function move(int $collectiveId, int $id, ?int $parentId, ?string $title,
 
 		try {
-			$this->updatePage($collectiveId, $file->getId(), $userId);
+			$this->updatePage($collectiveId, $file->getId(), $userId, slug: $slug);
```
changed

Koc commented on tests/stub.phpstub:
```diff
@@ -653,6 +653,20 @@ namespace OCA\Circles\Model\Probes {
 	}
 }
 
+namespace OCA\Circles\Model\Federated {
```
I've reordered them, if something is wrong - please say what code block should be where

Koc commented on lib/Migration/Version021200Date20240820000000.php:
```diff
@@ -0,0 +1,76 @@
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+class Version021200Date20240820000000 extends SimpleMigrationStep {
```
updated

Koc commented on psalm.xml:
```diff
@@ -31,6 +31,11 @@
 				<referencedClass name="OCP\SetupCheck\ISetupCheck" />
             </errorLevel>
         </UndefinedClass>
+		<UndefinedMethod>
```
moved to stubs

Koc commented:
hey @mejo- !

It seems like `./occ migrations:migrate collectives` works in a different way comparing to `./occ upgrade`. Here some interesting scenario:

1. Truncate all collectives/pages
2. Create few collectives with few pages. No need to delete anything.
3. Run `./occ migrations:migrate collectives`
4. Migrations migrated, slugs were successfully generated :heavy_check_mark:
5. Drop columns/migration records
6. Bump module version
7. Run `./occ upgrade` :boom: . There an error that collectives folder `/myCurrentUserName/files/Collectives/myCollectiveName` not exists

It doesn't matter have I removed collectives/pages or not. The fact is the same code works fine with `migrations:migrate` but not works with `upgrade`. I've try to debug it but no luck.

That's why I've moved slugs generation into separate console command. So, after upgrade old pages/collectives would have same urls like before.

<details><summary>Details</summary>
<p>

```
string(29) "/admin/files/Collectives/aaaa"  <---- exception message with path + collective name

#0 /var/www/html/lib/private/Files/Node/Folder.php(107): OC\Files\Node\Root->get()
#1 /var/www/html/apps/collectives/lib/Service/PageService.php(100): OC\Files\Node\Folder->get() <--- why collective folder not available? Only 1 page was trashed, not whole collective
#2 /var/www/html/apps/collectives/lib/Service/PageService.php(498): OCA\Collectives\Service\PageService->getCollectiveFolder()
#3 /var/www/html/apps/collectives/lib/Migration/Version021500Date20240820000001.php(78): OCA\Collectives\Service\PageService->findAll()
#4 /var/www/html/lib/private/DB/MigrationService.php(522): OCA\Collectives\Migration\Version021500Date20240820000001->postSchemaChange()
#5 /var/www/html/lib/private/DB/MigrationService.php(395): OC\DB\MigrationService->executeStep()
#6 /var/www/html/lib/private/legacy/OC_App.php(694): OC\DB\MigrationService->migrate()
#7 /var/www/html/lib/private/Updater.php(325): OC_App::updateApp()
#8 /var/www/html/lib/private/Updater.php(236): OC\Updater->doAppUpgrade()
#9 /var/www/html/lib/private/Updater.php(100): OC\Updater->doUpgrade()
#10 /var/www/html/core/Command/Upgrade.php(190): OC\Updater->upgrade()
#11 /var/www/html/3rdparty/symfony/console/Command/Command.php(326): OC\Core\Command\Upgrade->execute()
#12 /var/www/html/3rdparty/symfony/console/Application.php(1078): Symfony\Component\Console\Command\Command->run()
#13 /var/www/html/3rdparty/symfony/console/Application.php(324): Symfony\Component\Console\Application->doRunCommand()
#14 /var/www/html/3rdparty/symfony/console/Application.php(175): Symfony\Component\Console\Application->doRun()
#15 /var/www/html/lib/private/Console/Application.php(187): Symfony\Component\Console\Application->run()
#16 /var/www/html/console.php(87): OC\Console\Application->run()
#17 /var/www/html/occ(11): require_once('...')
#18 {main}
```

</p>
</details>

Koc commented:
I don't know why cypress is failing with `401: Unauthorized` error :confused:

juliusknorr commented:
Cypress failures were unrelated, should be fixed after a rebase to latest main by https://github.com/nextcloud/collectives/pull/1623

Koc commented:
@juliusknorr I've already rebased few times :smile:. Mentioned PR fixes Behat tests but not Cypress.

juliusknorr commented:
Ah sorry, seems I mixed up repos and have not pushed a fix for collectives yet. We'll need to adapt https://github.com/nextcloud/text/pull/6819 for this repo as well then

Koc commented:
@juliusknorr I've opened PR with backport: #1625.

But now there is another error :man_shrugging:  :

> `Could not download app contacts`

max-nextcloud commented:
@Koc I've finally found some time to look into this PR. Thanks a lot for starting it in the first place. I'd be happy to rebase it, resolve conflicts and take a look at cypress next week. Would this be okay or do you have local changes that you have not pushed yet? Don't want to step on your feet.

Koc commented:
@max-nextcloud I've just rebased branch and resolve conflicts. Yeah, I will be very appreciated if you can help me with Cypress. I've completely stuck with it. Feel free to push into my branch

max-nextcloud commented:
I'm prioritizing #1685 for now so we can start with a clean ci and rebase this on top of it. Sorry this is taking so long.

max-nextcloud commented:
I rebased on top of current master to resolve conflicts.

juliusknorr commented:
@mejo- Just came to my mind in the call with @max-nextcloud - This could also be a building block for removing the unique constraint on the collective name

max-nextcloud commented:
> @mejo- Just came to my mind in the call with @max-nextcloud - This could also be a building block for removing the unique constraint on the collective name

Currently we are limiting the collective name to not match the name of an existing team. So that would still be in effect even if the url schema changes.

max-nextcloud commented:
## Proposal: no collective id in urls

I propose to only use the collective name or the slug in the url. We could use one endpoint with a single parameter `:collective`. If we find a collective with a matchings slug we render it. If there's not collective with that slug we look for a collective with that name.

During the migration we generate slugs for all existing collectives, starting with collectives whose names could be a slug. We set their slugs to match the name. Then we generate slugs for the remaining collectives. If a slug has already been used we append a number (`slug-1`) similarly to how we handle file name conflicts.

This way we ensure that old urls still point to the same collective.

After the migration when rendering links ourselves we always use the slug. But we cannot update existing links so we also allow accessing collectives by their names.

When changing a slug or creating a new collective we follow the same principle. We can then later remove the restriction to require destinct names as the slugs may already have conflicts for distinct names such as 'hello' and 'hellö'.

Koc commented:
>  One benefit of including the ids is that the collective / page can still be identified after a rename.

Yeah, it was a main idea of having it as part of the url.

I did the suash & rebase with conflicts solving. But we still have failed Cypress tests which I not able to fix

Koc commented:
@max-nextcloud 

>  One benefit of including the ids is that the collective / page can still be identified after a rename.

Yeah, it was a main idea of having it as part of the url.

I did the suash & rebase with conflicts solving. But we still have failed Cypress tests which I not able to fix

max-nextcloud commented:
Thanks for the rebase, @Koc 
@mejo- will take another look at the migrations and I will take a stab at fixing the cypress tests.

mejo- commented on lib/Service/PageService.php:
```diff
@@ -763,8 +769,9 @@ public function copy(int $collectiveId, int $id, ?int $parentId, ?string $title,
+		$slug = $this->generateSlugForPage($title ?: $pageInfo->getTitle(), $file);
 		try {
-			$this->updatePage($collectiveId, $file->getId(), $userId, $pageInfo->getEmoji(), $pageInfo->isFullWidth());
+			$this->updatePage($collectiveId, $file->getId(), $userId, $pageInfo->getEmoji(), null, $slug);
```
```suggestion
			$this->updatePage($collectiveId, $file->getId(), $userId, $pageInfo->getEmoji(), $pageInfo->isFullWidth(), $slug);
```

I guess you removed `$pageInfo->isFullWidth()` by accident.

mejo- commented on lib/Service/PageService.php:
```diff
@@ -1101,4 +1110,12 @@ public function getBacklinks(int $collectiveId, int $id, string $userId): array
 	}
+
+	private function generateSlugForPage(string $title, ?File $file): ?string {
+		if (!$file) {
```
Can you explain why you check for `$file` here? I don't understand why this would be necessary.

Koc replied:
I don't remember why. Maybe previously there were cases when `$file` can be nullable, e.g. new page creation.

I've removed this condition and inlined method

mejo- commented on src/components/Collective.vue:
```diff
@@ -94,6 +97,20 @@ export default {
 			this.selectVersion(null)
+
+			const routerParams = this.$router.currentRoute.params
+			// If the current page is not the one we are supposed to be on, redirect
```
I have to admit that I don't fully understand the purpose of this code. When is this expected to happen? Is the idea to redirect to the slugified URL when the old URL is opened?

Koc replied:
yeah, exactly. The only one thing: it stopped work after one of latest rebases. I'm investigating it

mejo- commented on src/router.js:
```diff
@@ -16,22 +16,50 @@ const routes = [
+		component: CollectiveView,
+		props: (route) => route.params,
+		children: [
+			{ path: 'page-:pageId(\\d+)-:pageSlug' },
```
I think it's confusing to use `<slug>-<id>` for the collective URL part and `page-<id>-<slug>` for the page URL part. Why not use `<slug>-<id>` for the page as well?

Koc replied:
This was added as a prefix for some edge cases when generated slug/old file path contains digits and dash at the beginning

mejo- commented on src/stores/collectives.js:
```diff
@@ -41,11 +46,11 @@ export const useCollectivesStore = defineStore('collectives', {
 		collectivePath() {
 			return (collective) => {
 				const rootStore = useRootStore()
+				const slug = collective.slug ? collective.slug : encodeURIComponent(collective.name)
```
```suggestion
				const slug = collective.slug ??: encodeURIComponent(collective.name)
```

mejo- commented on src/stores/collectives.js:
```diff
@@ -76,7 +81,11 @@ export const useCollectivesStore = defineStore('collectives', {
+			if (!collective) {
+				return false
+			}
+			const slug = collective.slug ? collective.slug : encodeURIComponent(collective.name)
```
```suggestion
			const slug = collective.slug ?? encodeURIComponent(collective.name)
```

mejo- commented:
> @juliusknorr I've opened PR with backport: #1625.
> 
> But now there is another error 🤷‍♂️ :
> 
> > `Could not download app contacts`

@Koc seems like the CI job runs into a rate limit of the app store. I wonder whether this is because the base branch for this PR lives in your Github realm and not in `nextcloud`. Given that you're member of the `nextcloud` organisation, would you mind opening the PR again as a branch of `nextcloud/collectives`, instead of `Koc/nextcloud-collectives`? I can imagine that this helps with the failing Cypress tests.

Koc commented:
hey @mejo- !

I will try to do next things: 
- fix/answer to your code review comments :construction:
- fix conflicts after rebase :heavy_check_mark:
- do manual testing :hourglass_flowing_sand: 
- reopen new PR based on main repository instead of my fork :hourglass:

And after that I will disappear :smile: . I don't know how to fix Cypress tests

mejo- commented:
> fix conflicts after rebase ✔️

Sorry @Koc, I just merged another larger PR that created rebase/merge conflicts for this PR one more time. They should be easy to fix this time though. Maybe you can open a new PR based on a branch in this repo soon, then I can do the manual rebase myself next time :face_with_peeking_eye:

Koc commented on lib/Service/SlugService.php:
```diff
@@ -0,0 +1,27 @@
+		return $this->slugger->slug($name)->toString() . '-' . $collectiveId;
+	}
+
+	public function generatePageSlug(string $title): string {
```
no, we need ids for both pages and collectives. This is needed because collective/page can be renamed but we want to make it possible content using old URLs.

At the same time I can't remember why we can't store fileId inside slug on BE side (PR was opened almost year ago). I guess it's because of routing stuff - we need to distinguish old and new routes

![image](https://github.com/user-attachments/assets/a40cb4dc-0b26-4482-92fe-8c14e064f441)

Koc commented:
So, here we go. I did rebase one more time and fixed few things.
Here my manual test scenario:
1. switch to the current `main` branch
2. install module `./occ app:enable circles text collectives`
3. create 2 collectives with some nested pages
4. switch to my branch
5. run `./occ upgrade`
6. open collectives and pages by legacy urls -> it stil works, navigation is possible if slugs not generated :heavy_check_mark: 
7. run `./occ maintenance:repair` -> slugs generated and present in database for all collectives/pages :heavy_check_mark: 
    <details><summary>:mag: Preview</summary>
    <p>

   ![image](https://github.com/user-attachments/assets/7bf2ab12-ca30-4d07-93a1-3793a54fca57)

    </p>
    </details> 
9. refresh page -> redirected to new slugified url, validated for both collective index/pages :heavy_check_mark: 
10. rename collective (actually circle) -> slug changes, we've got redirected to the new url after change :heavy_check_mark: 

Here reopened PR #1802 created from original Nextcloud repo instead of my own fork. Please don't abandon it  :pray: , I spent really too much time on such functionality

![image](https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExeXBnanQ4bWJ5NTZocWE3Nzh2ZzFqZm9pbTE2OTdsbmsxN3QzMDhkeSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/3o7qDSOvfaCO9b3MlO/giphy.gif)

cc @mejo-
